### PR TITLE
[Android,iOS] Add Visual States to CarouselView items

### LINKED ISSUE #7904: [Bug] CarouselView doesn't implement the defined visual states
### Description

In the 4.3 branch, the CarouselView class defines the following visual states:

```
		public const string CurrentItemVisualState = "CurrentItem";
		public const string NextItemVisualState = "NextItem";
		public const string PreviousItemVisualState = "PreviousItem";
		public const string VisibleItemVisualState = "VisibleItem";
		public const string DefaultItemVisualState = "DefaultItem";
```

I spent a couple of hours trying to get the damn things to work and the only one I had success with was the DefaultItem state.

After staring at the code some more I discovered that only the DefaultItem state is implemented, and that's iOS only. The rest aren't implemented at all.

### LINKED ISSUE #7997: [Bug] CarouselView VisibleViews property is always null
### Description

In a `CarouselView` the `VisibleViews` property is always `null`. This is because there's no code in the 4.3 branch that ever sets the property.

### LINKED ISSUE #8671: [Bug] CarouselView - CurrentItem is not binded properly when switching pages programmatically
I have a setup where the carouselview has a binded property "CurrentItem".  When we set currentitem to the next page programatically the binded property will be set 4 times.  3 times the proper item, but 1 time it is the wrong (previous) object. 
This has a negative effect on components that respond to the current page index like page button states or a stepsindicator.

**Xaml:** 

```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:d="http://xamarin.com/schemas/2014/forms/design"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d"
             x:Class="TestCarousel.MainPage">
    <StackLayout>
        <StackLayout VerticalOptions="FillAndExpand" HorizontalOptions="FillAndExpand">
            <CarouselView 
                x:Name="testCarousel"
                ItemsSource="{Binding TestCollection}"
                CurrentItem="{Binding CurrentTestModel}">
                <CarouselView.ItemTemplate>
                    <DataTemplate>
                        <StackLayout>
                            <Label Text="Title"/>
                            <Label Text="{Binding Title}"/>
                            <Label Text="Description"/>
                            <Label Text="{Binding Description}"/>
                        </StackLayout>
                    </DataTemplate>
                </CarouselView.ItemTemplate>
            </CarouselView>
            <Button Text="Previous" Command="{Binding PreviousPageCommand}" />
            <Button Text="Next" Command="{Binding NextPageCommand}" />
        </StackLayout>
    </StackLayout>
</ContentPage>
```

**codebehind:**

```
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using Xamarin.Forms;

namespace TestCarousel
{
    // Learn more about making custom code visible in the Xamarin.Forms previewer
    // by visiting https://aka.ms/xamarinforms-previewer
    [DesignTimeVisible(false)]
    public partial class MainPage : ContentPage, INotifyPropertyChanged
    {
        public MainPage()
        {
            InitializeComponent();

            this.Appearing += MainPage_Appearing;
            this.BindingContext = this;

            this.testCarousel.Scrolled += TestCarousel_Scrolled;
            this.testCarousel.CurrentItemChanged += TestCarousel_CurrentItemChanged;
        }

        private void MainPage_Appearing(object sender, EventArgs e)
        {
            this.TestCollection = new ObservableCollection<TestModel>()
            {
                new TestModel() { Title = "Title 1", Description = "Description 1"},
                new TestModel() { Title = "Title 2", Description = "Description 2"},
                new TestModel() { Title = "Title 3", Description = "Description 3"},
                new TestModel() { Title = "Title 4", Description = "Description 4"},
                new TestModel() { Title = "Title 5", Description = "Description 5"},
            };

            this.CurrentTestModel = TestCollection[0];
        }

        private ObservableCollection<TestModel> testCollection;
        public ObservableCollection<TestModel> TestCollection 
        {
            get
            {
                return testCollection;
            }
            set
            {
                testCollection = value;
                RaisePropertyChanged();
            }
        }

        private int pageIndex = 0;
        public int PageIndex
        {
            get
            {
                return pageIndex;
            }
            set
            {
                pageIndex = value;
            }
        }

        private TestModel currentTestModel;
        public TestModel CurrentTestModel 
        {
            get
            {
                return currentTestModel;
            }
            set
            {
                currentTestModel = value;
                Debug.WriteLine($"CurrentPage changed to {TestCollection.IndexOf(value)}");
                RaisePropertyChanged();
            }
        }


        public Command NextPageCommand
        {
            get
            {
                return new Command(() => {
                    PageIndex++;
                    CurrentTestModel = TestCollection[PageIndex];
                });
            }
        }

        public Command PreviousPageCommand
        {
            get
            {
                return new Command(() => {
                    PageIndex--;
                    CurrentTestModel = TestCollection[PageIndex];
                });
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;

        /// <summary>
        /// Raises the property changed.
        /// </summary>
        /// <param name="propertyName">(Optional) Property name.</param>
        protected void RaisePropertyChanged([CallerMemberName] string propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }
    }

    public class TestModel
    {
        public string Title { get; set; }
        public string Description { get; set; }
    }
}
```

**Steps to reproduce:**

Unzip the sample project in the zipfile and run it. 

[CarouselView - Bugreport.zip](https://github.com/xamarin/Xamarin.Forms.CarouselView/files/3804815/CarouselView.-.Bugreport.zip)

Tab the next button and you will see the following debugging line: 

image

Results in the following debugging output:

image

**System info:**

Xamarin.Forms: 4.3.0.947036
Tested on: Android 8, Android 9

**Update:**

I have tested the same code in iOS (build against xcode 11) and it results in the same kind of problems. 

When you tap on the next button, carousel does not animate to the next page. In XF 4.3.0.819712-pre2  it works perfectly fine but in the latest stable (4.3.0.947036) it is not working

Tested on iPhone 6 - iOS 12.4

### PR DESCRIPTION
**Pull request by rmarinho**
### Description of Change ###

Implement the visual states changes on Carouselview 

### Issues Resolved ### 

- fixes #7904
- fixes #7997  
- fixes #8671

### API Changes ###

 None

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android

### Behavioral/Visual Changes ###

Setting diffrent properties using VisualStateManager should work

### Testing Procedure ###

Check the xaml example on CarouselView gallery, you should see the previous and next item smaller and with 0.7  opacity , and the current item should be the correct full size 

### PR Checklist ###

- [x] Targets the correct branch
- [x] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Controls/GalleryPages/CollectionViewGalleries/CarouselViewGalleries/ExampleTemplateCarousel.xaml.cs [hunk 1]
```diff
@@ -56,7 +58,15 @@ public ExampleTemplateCarousel()
 					}
 				}
 			};
-			GestureRecognizers.Add(gesture);
+			//GestureRecognizers.Add(gesture);
```
Comment (rmarinho) [inline Xamarin.Forms.Controls/GalleryPages/CollectionViewGalleries/CarouselViewGalleries/ExampleTemplateCarousel.xaml.cs]: Adding a gesture to the Item template breaks scrolling.

File: Xamarin.Forms.Core/Items/CarouselView.cs [hunk 2]
```diff
@@ -161,6 +171,12 @@ public LinearItemsLayout ItemsLayout
 		[EditorBrowsable(EditorBrowsableState.Never)]
 		public bool IsScrolling { get; set; }
 
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public bool IsInitialized = false;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Items/CarouselView.cs]: name should be more explicit. it should also be changed to a property or a method, as it's value is set from outside of this type

Comment (rmarinho) [reply Xamarin.Forms.Core/Items/CarouselView.cs]: Moved to a method  a new name

File: Xamarin.Forms.Controls/GalleryPages/CollectionViewGalleries/CarouselViewGalleries/ExampleTemplateCarousel.xaml.cs (hunk 1 above)
Comment (samhouts) [reply Xamarin.Forms.Controls/GalleryPages/CollectionViewGalleries/CarouselViewGalleries/ExampleTemplateCarousel.xaml.cs]: Do we have an open issue for that yet?

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 3]
```diff
@@ -7,21 +9,34 @@
 using Android.Support.V7.Widget;
 #endif
 using Android.Views;
+using Java.Interop;
 using FormsCarouselView = Xamarin.Forms.CarouselView;
 
 namespace Xamarin.Forms.Platform.Android
 {
+	class CarrouselViewwOnGlobalLayoutListener : Java.Lang.Object, ViewTreeObserver.IOnGlobalLayoutListener
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
	class CarouselViewOnGlobalLayoutListener : Java.Lang.Object, ViewTreeObserver.IOnGlobalLayoutListener
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 4]
```diff
@@ -7,21 +9,34 @@
...
+
 	public class CarouselViewRenderer : ItemsViewRenderer<ItemsView, ItemsViewAdapter<ItemsView, IItemsViewSource>, IItemsViewSource>
 	{
 		protected FormsCarouselView Carousel;
 		RecyclerView.ItemDecoration _itemDecoration;
 		bool _isSwipeEnabled;
 		int _oldPosition;
 		int _initialPosition;
+		List<View> _oldViews;
+		CarrouselViewwOnGlobalLayoutListener _carrouselViewLayoutListener;
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
		CarouselViewOnGlobalLayoutListener _carouselViewLayoutListener;
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 5]
```diff
@@ -212,7 +247,105 @@ void UpdateInitialPosition()
...
+		}
+
+		void CarouselViewScrolled(object sender, ItemsViewScrolledEventArgs e)
+		{
+			UpdateVisualStates();
+		}
+
+		void AddLayoutListener()
+		{
+			_carrouselViewLayoutListener = new CarrouselViewwOnGlobalLayoutListener();
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
			_carouselViewLayoutListener = new CarouselViewOnGlobalLayoutListener();
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 6]
```diff
@@ -212,7 +247,105 @@ void UpdateInitialPosition()
...
+			_carrouselViewLayoutListener.LayoutReady += LayoutReady;
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
			_carouselViewLayoutListener.LayoutReady += LayoutReady;
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 7]
```diff
@@ -212,7 +247,105 @@ void UpdateInitialPosition()
...
+
+			ViewTreeObserver.AddOnGlobalLayoutListener(_carrouselViewLayoutListener);
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
			ViewTreeObserver.AddOnGlobalLayoutListener(_carouselViewLayoutListener);
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 8]
```diff
@@ -212,7 +247,105 @@ void UpdateInitialPosition()
...
+			}
+
+			Carousel.PlatformInitialized();
+			UpdateVisualStates();
+			ClearLayoutListener();
+		}
+
+		void ClearLayoutListener()
+		{
+			if (_carrouselViewLayoutListener == null)
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
			if (_carouselViewLayoutListener == null)
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 9]
```diff
@@ -212,7 +247,105 @@ void UpdateInitialPosition()
...
+				return;
+
+			ViewTreeObserver?.RemoveOnGlobalLayoutListener(_carrouselViewLayoutListener);
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
			ViewTreeObserver?.RemoveOnGlobalLayoutListener(_carouselViewLayoutListener);
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 10]
```diff
@@ -212,7 +247,105 @@ void UpdateInitialPosition()
...
+			_carrouselViewLayoutListener.LayoutReady -= LayoutReady;
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
			_carouselViewLayoutListener.LayoutReady -= LayoutReady;
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 11]
```diff
@@ -212,7 +247,105 @@ void UpdateInitialPosition()
...
+			_carrouselViewLayoutListener = null;
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]:
```suggestion
			_carouselViewLayoutListener = null;
```

File: Xamarin.Forms.Platform.iOS/CollectionView/CarouselViewController.cs [hunk 12]
```diff
@@ -8,12 +9,21 @@ public class CarouselViewController : ItemsViewController<CarouselView>
...
 		{
 			_carouselView = itemsView;
 			CollectionView.AllowsSelection = false;
 			CollectionView.AllowsMultipleSelection = false;
+			_carouselView.PropertyChanged += CarouselViewPropertyChanged;
+			_carouselView.Scrolled += CarouselViewScrolled;
+			_oldViews = new List<View>();
+		}
+
+		private void CarouselViewScrolled(object sender, ItemsViewScrolledEventArgs e)
```
Comment (samhouts) [inline Xamarin.Forms.Platform.iOS/CollectionView/CarouselViewController.cs]:
```suggestion
		void CarouselViewScrolled(object sender, ItemsViewScrolledEventArgs e)
```

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs (hunk 3 above)
Comment (rmarinho) [reply Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]: Dahh carrousel is portuguese :p

**Pull request merged**